### PR TITLE
[Mercure] update mercure documentation

### PR DESCRIPTION
**Pull request by azjezz**

### CODE REVIEW

File: mercure.rst [hunk 1]
```diff
@@ -159,20 +159,26 @@ service, including controllers::
...
-    use Symfony\Component\Mercure\PublisherInterface;
+    use Symfony\Component\Mercure\HubInterface;
     use Symfony\Component\Mercure\Update;
 
     class PublishController
     {
-        public function __invoke(PublisherInterface $publisher): Response
+        private $hub;
+
+        public function __construct(HubInterface $hub)
```
Comment (chalasr) [inline mercure.rst]: I suggest to stick to controller argument injection for simplicity and consistency with the rest of the docs.

File: mercure.rst [hunk 2]
```diff
@@ -417,39 +432,34 @@ And here is the controller::
...
+    use Symfony\Component\Mercure\Authorization;
+    use Symfony\Component\Mercure\Discovery;
 
-    class DiscoverController extends AbstractController
+    class DiscoverController
     {
+        private $discovery;
+        private $authorization;
+
+        public function __construct(Discovery $discovery, Authorization $authorization)
```
Comment (dunglas) [inline mercure.rst]: Same here, you can use controller injection for consistency.

File: mercure.rst [hunk 3]
```diff
@@ -417,39 +432,34 @@ And here is the controller::
...
-                ->withValue($token)
-                ->withPath('/.well-known/mercure')
-                ->withSecure(true)
-                ->withHttpOnly(true)
-                ->withSameSite('strict')
-            ;
-            $response->headers->setCookie($cookie);
+            $this->discovery->addLink($request);
+
+            $response = new JsonResponse([
```
Comment (dunglas) [inline mercure.rst:463]: I suggest using the `$this->json()` method provided by `AbstractController` for consistency too.

File: mercure.rst [hunk 4]
```diff
@@ -464,15 +474,17 @@ Programmatically Generating The JWT Used to Publish
 ---------------------------------------------------
 
 Instead of directly storing a JWT in the configuration,
-you can create a service that will return the token used by
-the ``Publisher`` object::
+you can create a token provider that will return the token used by
```
Comment (dunglas) [inline mercure.rst:485]: Could you also document the new `LcobucciFactory` class?

**Review by dunglas — CHANGES_REQUESTED** [review]
I tagged the new releases, we can merge this one now! I made some minor change requests.

**Comment by javiereguiluz** [comment]
In which version should we merge this? This article is available in 4.4, 5.2 and 5.x branches. Thanks!

**Comment by azjezz** [comment]
> In which version should we merge this? This article is available in 4.4, 5.2 and 5.x branches. Thanks!

mercure and mercure-bundle don't follow symfony versioning, these changes apply to mercure 0.5 and mercure-bundle 0.3, so i'm not really sure.

File: mercure.rst [hunk 5]
```diff
@@ -412,44 +408,72 @@ To generate the JWT, we'll use the ``lcobucci/jwt`` library. Install it:
 
     $ composer require lcobucci/jwt
```
Comment (chalasr) [inline mercure.rst]: it's no longer needed

File: mercure.rst [hunk 6]
```diff
@@ -573,29 +603,59 @@ its Mercure support.
...
+    use App\Controller\MessageController;
+    use Symfony\Component\Mercure\HubInterface;
+    use Symfony\Component\Mercure\JWT\StaticTokenProvider;
+    use Symfony\Component\Mercure\MockHub;
     use Symfony\Component\Mercure\Update;
 
-    class PublisherStub implements PublisherInterface
+    class MessageControllerTest extends TestCase
     {
         public function __invoke(Update $update): string
```
Comment (chalasr) [inline mercure.rst]: the signature does not look right

**Comment by chalasr** [comment]
As symfony/mercure and mercure-bundle are still in 0.x versions, and because the latest versions support all maintained Symfony versions, I suggest to merge this in the 4.4 branch.

**Comment by javiereguiluz** [comment]
@azjezz thanks for this contribution! And thanks to reviewers too!

This is now merged in 4.4 branch and up.

**Pull request merged**